File: CodeChef/DecemberChallenge2019/binary_xor.cpp
#include<bits/stdc++.h>
#define F first
#define S second
#define pb push_back
#define mp make_pair
#define full(a) a.begin(),a.end()
#define rfull(a) a.rbegin(),a.rend()

using namespace std;
const double pi=acos(-1.0);
const double pii=2*pi;
const double eps=1e-6;
const long long MOD=1e9+7;

long long pow_mod(long long a, long long b){
	if(a == 0)
		return 0ll;
	if(b == 0)
		return 1ll;
	if(b == 1)
		return a%MOD;
	long long x = pow_mod(a,b/2);
	x = (x*x)%MOD;
	if(b%2==1)
		return (x*a)%MOD;
	return x;
}

long long nCr(int n, int r, vector<long long>& fact){
	if(n<0 || r<0 || n<r)
		return 0ll;
	return (fact[n]*pow_mod((fact[n-r]*fact[r])%MOD,MOD-2))%MOD;
}

void solve(){
	int n;
	cin>>n;
	string str_a,str_b;
	vector<long long> factorial_i(n+1,1ll);
	for(int i=1;i<=n;i++)
		factorial_i[i]=(factorial_i[i-1]*i)%MOD;
	cin>>str_a>>str_b;
	int x=0,y=0;
	for(int i=0;i<n;i++){
		x+=(str_a[i]=='1' ? 1:0);
		y+=(str_b[i]=='1' ? 1:0);
	}

	if(true){
		int t1,t2;
		t1=min(x,y);
		t2=min(n-x,n-y);
		x=t1;
		y=t2;
	}
	long long ans=0;
	for(int i=0;i<=min(x,y);i++)
		ans=(ans+nCr(n,n-x-y+2*i,factorial_i))%MOD;
	cout<<ans<<"\n";
}

int main(){
	std::ios::sync_with_stdio(false);
	std::cin.tie(NULL);
	std::cout.tie(NULL);

	int t;       
	cin >> t;
	while(t--){
		solve();
	}
	return 0;
}

File: CodeChef/DecemberChallenge2019/binary_xor.py
def pow_mod(a,b,mod):
	if a == 0:
		return 0
	if b == 0:
		return 1
	if b == 1:
		return a%mod

	x = pow_mod(a,b//2,mod)%mod
	x = (x*x)%mod
	if b%2 == 1:
		return (x*a)%mod
	return x

def ncr(n,i,num,den,mod):
	if i == 0 or i == n:
		return 1
	den = pow_mod(den,mod-2,mod)%mod
	return (num*den)%mod

tc = int(input())
ans_str=""
for t in range(tc):
	n = int(input())
	a = input()
	b = input()

	x1,x2,y1,y2=0,0,0,0
	ans,mod = 0,int(1e9)+7
	for i in range(n):
		x1 += (1 if a[i] == '0' else 0)
		x2 += (1 if b[i] == '0' else 0)
	y1=n-x1
	y2=n-x2
	pair_zero = min(x1,x2)
	pair_one = min(y1,y2)
	# print(max_pair,max_non_pair)
	nfact_Array = [1 for i in range(n+1)]
	for i in range(1,len(nfact_Array)):
		nfact_Array[i] = (nfact_Array[i-1]*i)%mod

	flag = True
	while flag:
		i = n-(pair_one + pair_zero)
		if i >= 0 and i <= n+1:
			ans += ncr(n,i,nfact_Array[n],(nfact_Array[i]*nfact_Array[n-i])%mod,mod)
			ans = ans%mod
		if pair_zero > 0 and pair_one > 0:
			pair_one -= 1
			pair_zero -= 1
		else:
			flag = False

	ans_str = ans_str + str(ans) + "\n"

print(ans_str,end="")

